delete default css

ralfiannor opened the pull request:
Implement #253

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/livebud/bud?pullRequest=254) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/livebud/bud?pullRequest=254) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/livebud/bud?pullRequest=254) it.</sub>

matthewmueller commented:
Thanks for the PR @ralfiannor and for kicking this off! I don't want to remove the default.css without providing an alternative solution. I've wrote my thoughts up in #253. Let me know if you want to drive one of those solutions forward or if you have any other ideas!

matthewmueller commented on framework/view/ssr/svelte.ts:
```diff
@@ -58,7 +58,7 @@ const defaultLayout = {
           <head>
             <meta charset="utf-8"/>
-            <style>${defaultCSS}</style>
+			<link rel="stylesheet" href="default.css">
```
Formatting

matthewmueller commented on framework/view/ssr/svelte.js:
```diff
@@ -327,7 +327,7 @@ var defaultLayout = {
           <head>
             <meta charset="utf-8"/>
-            <style>${defaultCSS}</style>
+      			<link rel="stylesheet" href="default.css">
```
Spacing looks off

matthewmueller commented on internal/cli/create/create_test.go:
```diff
@@ -38,6 +38,7 @@ func TestCreateOutsideGoPath(t *testing.T) {
 	is.NoErr(td.Exists("go.sum"))
 	is.NoErr(td.Exists("package.json"))
 	is.NoErr(td.Exists("package-lock.json"))
+	is.NoErr(td.Exists("public/default.css"))
```
I think we're missing some test cases still:
1. The scaffolded `public/default.css` gets deleted but views and public files still work as expected.
2. Someone brings their own `public/default.css` and that gets picked up and used.

ralfiannor replied:
Added the test case in `public_test.go`

matthewmueller commented on internal/embedded/embedded.go:
```diff
@@ -7,7 +7,19 @@ import (
 }
+
+// DefaultCss returns the default css data
+func DefaultCss(css string) string {
```
I think we can just use function names like `func Normalize() []byte` or `func Reset() []byte` since it's pretty clear where the context comes from.

matthewmueller reviewed (commented):
Thanks for working on this! I'd just like us to clean a few things up and add a few more tests (maybe in framework/public/public_test.go) and then I think we're good to go!

matthewmueller commented on framework/controller/controller_test.go:
```diff
@@ -1285,7 +1285,6 @@ func TestViewUnnamed(t *testing.T) {
 	is.NoErr(err)
 	is.NoErr(res.DiffHeaders(`
 		HTTP/1.1 200 OK
-		Transfer-Encoding: chunked
```
I'm quite surprised the changes affected these tests. Do you know why you needed to remove this?

ralfiannor replied:
Reverted. But I still get an error on it while running the unit test. I use `go 1.7` and `node 18.9`

matthewmueller commented on internal/embedded/embedded.go:
```diff
@@ -7,7 +7,20 @@ import (
+// EmptyCss reset the default css data
+func EmptyCss() []byte {
+	return []byte("/* No Default CSS Loaded */")
+}
```
It'd be better if the view didn't include the `<link />` if a `public/default.css` isn't present since that would avoid a roundtrip for this CSS sheet. 

Then we can also remove this function.

ralfiannor replied:
Done

matthewmueller commented on internal/cli/cli.go:
```diff
@@ -49,6 +49,7 @@ func (c *CLI) Run(ctx context.Context, args ...string) error {
 		cmd := create.New(cmd, c.in)
 		cli := cli.Command("create", "create a new app")
 		cli.Arg("dir").String(&cmd.Dir)
+		cli.Flag("css", "add a css").String(&cmd.Css).Default("normalize")
```
Let's remove this flag for now, I'm not ready to expose it just yet, especially since it's not really possible to select an alternative.

ralfiannor replied:
Done

matthewmueller commented on framework/public/public_test.go:
```diff
@@ -144,8 +149,20 @@ func TestGetChangeGet(t *testing.T) {
+	res, err = app.Get("/default.css")
+	is.NoErr(err)
+	is.Equal(200, res.Status())
+	is.Equal(res.Body().Bytes(), defaultCss2)
```
It's a bit hard to follow these changes because you've added new conditions to existing tests. 

Let's try to only touch existing tests that need to be updated for the new behavior, then add new tests for the new behavior.

ralfiannor replied:
Reverted. No need to update on `public.go` just update on `ssr.go`

matthewmueller commented on framework/view/ssr/svelte.js:
```diff
@@ -327,7 +327,7 @@ var defaultLayout = {
           <head>
             <meta charset="utf-8"/>
-            <style>${defaultCSS}</style>
+            <link rel="stylesheet" href="default.css" />
```
Do you mind trying to feed the configuration through so you can remove this if `public/default.css` doesn't exist? Let me know if it's difficult and I'll give it a go in this PR.

ralfiannor replied:
Will try. I will add the checking file existing in `ssr.go` but will call our logic each page loaded. Is it okay ?

matthewmueller replied:
I think it should be done at compile time actually

matthewmueller reviewed (changes requested):
Just reviewed. This is getting closer @ralfiannor!

matthewmueller commented:
I just realized we should have the same treatment for the public/favicon.ico: scaffold a default one, but then give users the option to override it or delete it.

ralfiannor commented:
> I just realized we should also do the same thing for `public/favicon.ico`: scaffold a default one, but then give users the option to override it or delete it.

Can we do it on another PR?

matthewmueller commented:
Thanks for continuing to push this forward @ralfiannor! 

Do you know why Go is adding/removing `Transfer-Encoding: chunked`? 

Let's just remove the `Transfer-Encoding: chunked` in the view tests and I think this is good to go!

ralfiannor commented:
> Do you happen to know the condition in which Go is adding/removing `Transfer-Encoding: chunked`?

Yes, it happen when we added <style></style> the chunked also added

matthewmueller commented:
Thanks @ralfiannor! I just pulled it down and did some manual testing. Two issues popped up:

1. Welcome page no longer shows up when creating a new project
2. Renaming or removing the `default.css` requires you to restart the `bud` process

Adjusting the global `svelteRuntime` is a bit hacky. If you're up for cleaning this up, a better solution would be to pass this information through the `svelte.gotext` template and then reference that information in the `svelte.ts` file. I'm okay with cleaning this up later if you're not sure how to do this.

Here's a diff with the fixes:

```diff
diff --git a/framework/app/app_test.go b/framework/app/app_test.go
index b5ba644..500a117 100644
--- a/framework/app/app_test.go
+++ b/framework/app/app_test.go
@@ -30,3 +30,26 @@ func TestWelcome(t *testing.T) {
 	is.NoErr(td.Exists("bud/app"))
 	is.NoErr(app.Close())
 }
+
+func TestWelcomeWithPublic(t *testing.T) {
+	is := is.New(t)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	dir := t.TempDir()
+	td := testdir.New(dir)
+	td.Files["public/default.css"] = `* { box-sizing: border-box; }`
+	is.NoErr(td.Write(ctx))
+	cli := testcli.New(dir)
+	is.NoErr(td.NotExists("bud/app"))
+	app, err := cli.Start(ctx, "run")
+	is.NoErr(err)
+	res, err := app.Get("/")
+	is.NoErr(err)
+	is.Equal(res.Status(), 200)
+	is.In(res.Body().String(), "Hey Bud")
+	is.In(res.Body().String(), "Hey Bud") // should work multiple times
+	is.Equal(app.Stdout(), "")
+	is.Equal(app.Stderr(), "")
+	is.NoErr(td.Exists("bud/app"))
+	is.NoErr(app.Close())
+}
diff --git a/framework/view/ssr/ssr.go b/framework/view/ssr/ssr.go
index 7116bcd..8bab21c 100644
--- a/framework/view/ssr/ssr.go
+++ b/framework/view/ssr/ssr.go
@@ -50,9 +50,10 @@ type Compiler struct {
 
 func (c *Compiler) Compile(ctx context.Context, fsys budfs.FS) ([]byte, error) {
 	dir := c.module.Directory()
-
 	if existCss := vfs.Exist(fsys, "public/default.css"); nil == existCss {
 		svelteRuntime = strings.Replace(svelteRuntime, `<!-- default css -->`, `<link rel="stylesheet" href="default.css">`, 1)
+	} else {
+		svelteRuntime = strings.Replace(svelteRuntime, `<link rel="stylesheet" href="default.css">`, `<!-- default css -->`, 1)
 	}
 
 	result := esbuild.Build(esbuild.BuildOptions{
diff --git a/framework/web/loader.go b/framework/web/loader.go
index 5f0d149..e4695b8 100644
--- a/framework/web/loader.go
+++ b/framework/web/loader.go
@@ -52,7 +52,10 @@ func (l *loader) Load() (state *State, err error) {
 	l.imports.AddNamed("webrt", "github.com/livebud/bud/framework/web/webrt")
 	l.imports.AddNamed("router", "github.com/livebud/bud/package/router")
 	// Show the welcome page if we don't have controllers, views or public files
-	if len(exist) == 0 {
+	//
+	// TODO: welcome should be part of the default controller, then we can support
+	// a simple static file server without generating the welcome page.
+	if len(exist) == 0 || (len(exist) == 1 && exist["bud/internal/app/public/public.go"]) {
 		l.imports.AddNamed("welcome", "github.com/livebud/bud/framework/web/welcome")
 		state.ShowWelcome = true
 		state.Imports = l.imports.List()

```

matthewmueller commented:
Any chance of wrapping this up @ralfiannor ?
